### PR TITLE
[WIP] new test type design

### PR DESCRIPTION
**Pull request by bbguimaraes**
First prototype of the new test type. This version is capable of reproducing the most common template test (AWS E2E). See the corresponding WIP branch [here](https://github.com/bbguimaraes/release/tree/steps).

Some parts of the code are still specific to E2E tests, but I've isolated them in the last commit and am open to suggestions on how to make them more generic.

Other than that, next steps are (sup)porting more test types, adding more tests (which we can finally have), and making sure we have 100% feature parity with templates.

### CODE REVIEW

File: pkg/steps/new_e2e.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		}
+		name := s.name + "-" + step.Name
+		pod, err := generateBasePod(s.jobSpec, resources)
+		pod.ObjectMeta.Name = name
+		pod.ObjectMeta.Annotations[annotationContainersForSubTestResults] = step.Name
+		container := &pod.Spec.Containers[0]
+		container.Name = step.Name
+		container.Image = step.Image
+		container.Command = []string{"/bin/sh", "-c", "#!/bin/sh\nset -eu\n" + step.Commands}
+		container.Env = append(container.Env, []coreapi.EnvVar{
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: I think this has changed (perhaps since you last looked) -- look for Hongkai's commit to see how to add all env vars

File: pkg/steps/new_e2e.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	if errs != nil {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	return ret, nil
+}
+
+func (s newE2ETestStep) runPods(ctx context.Context, pods []coreapi.Pod) error {
+	go func() {
+		<-ctx.Done()
+		for _, pod := range pods {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: Could do this with a label selector too, to make for one call

File: pkg/steps/new_e2e.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+			}
+		}
+	}()
+	for _, pod := range pods {
+		log.Printf("Executing %q", pod.Name)
+		if _, err := createOrRestartPod(s.podClient.Pods(s.jobSpec.Namespace), &pod); err != nil {
+			return fmt.Errorf("failed to create or restart %q pod: %v", pod.Name, err)
+		}
+		if err := waitForPodCompletion(s.podClient.Pods(s.jobSpec.Namespace), pod.Name, nil, false); err != nil {
+			return fmt.Errorf("%q pod %q failed: %v", s.name, pod.Name, err)
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: Do we want, for instance, on cleanup tasks after the test has run, for all steps to try to run regardless?

File: pkg/steps/pod.go [hunk 4]
```diff
@@ -193,42 +193,48 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
...
-	}
+	}, nil
+}
 
+func (s *podStep) generatePodForStep(image string, containerResources coreapi.ResourceRequirements) (*coreapi.Pod, error) {
+	pod, err := generateBasePod(s.jobSpec, containerResources)
+	if err != nil {
+		return nil, err
+	}
+	pod.ObjectMeta.Name = s.config.As
```
Comment (stevekuznetsov) [inline pkg/steps/pod.go]: All of these look like the same post-generation edits as in the other file. Why not pass them in or otherwise DRY them out?

File: pkg/steps/new_e2e_test.go [hunk 5]
```diff
@@ -13,10 +13,27 @@ import (
...
+func (p testParameters) Get(name string) (string, error)  { return p.m[name], nil }
+func (p testParameters) Links(name string) []api.StepLink { return nil }
+
 func TestGenPods(t *testing.T) {
 	config := api.TestStepConfiguration{
 		As: "test",
 	}
+	params := testParameters{
+		m: map[string]string{
+			"IMAGE_IMAGE0": "registry.example.com/test/image0",
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e_test.go]: Why didn't we see this affect the expected test output?

File: pkg/api/types.go [hunk 6]
```diff
@@ -307,6 +310,14 @@ type TestStepConfiguration struct {
 	OpenshiftInstallerCustomTestImageClusterTestConfiguration *OpenshiftInstallerCustomTestImageClusterTestConfiguration `json:"openshift_installer_custom_test_image,omitempty"`
 }
 
+type TestStep struct {
+	Name        string               `json:"name,omitempty"`
+	Image       string               `json:"image,omitempty"`
+	Commands    string               `json:"commands,omitempty"`
+	ArtifactDir string               `json:"artifact_dir,omitempty"`
+	Resources   ResourceRequirements `json:"resources,omitempty"`
```
Comment (stevekuznetsov) [inline pkg/api/types.go]: Why not use the resource mapping we already have?

File: cmd/ci-operator/main.go [hunk 7]
```diff
@@ -268,7 +271,20 @@ func (o *options) Complete() error {
...
+	if o.stepLibraryPath != "" {
+		data, err := ioutil.ReadFile(o.stepLibraryPath)
+		if err != nil {
+			return fmt.Errorf("failed to read step library (%v)", err)
+		}
+		var l []api.TestStep
+		if err := yaml.Unmarshal(data, &l); err != nil {
+			return fmt.Errorf("failed to load step library (%v)", err)
+		}
+		o.stepLibrary = make(map[string]api.TestStep, len(l))
```
Comment (stevekuznetsov) [inline cmd/ci-operator/main.go]: If you want this data structure, lets just serialize it instead of having to do post-processing. What happens if two names collide?

File: pkg/steps/new_e2e.go [hunk 8]
```diff
@@ -139,12 +145,26 @@ func (s newE2ETestStep) genPods(steps []api.TestStep) ([]coreapi.Pod, error) {
 		if owner := s.jobSpec.Owner(); owner != nil {
 			pod.OwnerReferences = append(pod.OwnerReferences, *owner)
 		}
+		if step.ArtifactDir != "" {
+			if artifacts == nil {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: Having a possibly-nil worker is a bit awkward. Why not:

```go
type collectionRecord struct{ podName, containerName string }
var toCollect []collectionRecord

// ....

for _, step := range steps {
    if step.ArtifactDir != "" {
        // add the volume mount and container
        toCollect = append(toCollect, collectionRecord{podName: pod.Name, containerName: container.Name})
    }
}

if len(toCollect) > 0 {
    // create worker
    // return notifier
}
```

File: pkg/steps/new_e2e.go [hunk 9]
```diff
@@ -139,12 +145,26 @@ func (s newE2ETestStep) genPods(steps []api.TestStep) ([]coreapi.Pod, error) {
 		if owner := s.jobSpec.Owner(); owner != nil {
 			pod.OwnerReferences = append(pod.OwnerReferences, *owner)
 		}
+		if step.ArtifactDir != "" {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: Do we ever allow for an unset artifacts dir?

File: pkg/steps/new_e2e_test.go [hunk 10]
```diff
@@ -94,8 +94,8 @@ func TestGenPods(t *testing.T) {
 			}},
 		},
 	}}
-	step := newNewE2ETestStep(config, params, nil, &jobSpec, nil)
-	ret, err := step.genPods(steps)
+	step := newNewE2ETestStep(config, params, nil, "", &jobSpec, nil)
+	ret, _, err := step.genPods(steps)
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e_test.go]: Shouldn't we be testing the output of this?

File: pkg/steps/new_e2e_test.go [hunk 11]
```diff
@@ -94,8 +94,8 @@ func TestGenPods(t *testing.T) {
 			}},
 		},
 	}}
-	step := newNewE2ETestStep(config, params, nil, &jobSpec, nil)
-	ret, err := step.genPods(steps)
+	step := newNewE2ETestStep(config, params, nil, "", &jobSpec, nil)
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e_test.go]: How come adding artifacts didn't change any test output?

File: pkg/steps/new_e2e.go [hunk 12]
```diff
@@ -181,6 +199,50 @@ func imageForStep(image string, params api.Parameters) (string, error) {
 	return image, nil
 }
 
+func addSecretMount(pod *coreapi.Pod) {
+	path := "/tmp/installer"
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: this seems awfully specific

File: pkg/steps/new_e2e.go [hunk 13]
```diff
@@ -66,14 +75,20 @@ func (s *newE2ETestStep) Run(ctx context.Context, dry bool) error {
...
+	if secret, err = createSecret(s.secretClient, s.jobSpec.Namespace, s.artifactDir); err != nil {
+		errs = append(errs, fmt.Errorf("failed to create secret: %v", err))
+	}
+	if errs == nil {
+		if err = s.runSteps(ctx, s.tests, secret); err != nil {
 			errs = append(errs, fmt.Errorf("%q test steps failed: %v", s.name, err))
 		}
 	}
-	if err := s.runSteps(ctx, s.post); err != nil {
+	if err = s.runSteps(ctx, s.post, secret); err != nil {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: This should remain `err :=` ?

File: pkg/steps/new_e2e.go [hunk 14]
```diff
@@ -66,14 +75,20 @@ func (s *newE2ETestStep) Run(ctx context.Context, dry bool) error {
...
 		errs = append(errs, fmt.Errorf("%q pre steps failed: %v", s.name, err))
-	} else {
-		if err := s.runSteps(ctx, s.tests); err != nil {
+	}
+	var secret string
+	if secret, err = createSecret(s.secretClient, s.jobSpec.Namespace, s.artifactDir); err != nil {
+		errs = append(errs, fmt.Errorf("failed to create secret: %v", err))
+	}
+	if errs == nil {
+		if err = s.runSteps(ctx, s.tests, secret); err != nil {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: This should remain `err :=` ?

File: pkg/steps/new_e2e.go [hunk 15]
```diff
@@ -66,14 +75,20 @@ func (s *newE2ETestStep) Run(ctx context.Context, dry bool) error {
...
+	if err = s.runSteps(ctx, s.pre, ""); err != nil {
 		errs = append(errs, fmt.Errorf("%q pre steps failed: %v", s.name, err))
-	} else {
-		if err := s.runSteps(ctx, s.tests); err != nil {
+	}
+	var secret string
+	if secret, err = createSecret(s.secretClient, s.jobSpec.Namespace, s.artifactDir); err != nil {
+		errs = append(errs, fmt.Errorf("failed to create secret: %v", err))
+	}
+	if errs == nil {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: The logic here is getting a bit nuts. Why not just return when we can't make the secret?

File: pkg/steps/new_e2e.go [hunk 16]
```diff
@@ -66,14 +75,20 @@ func (s *newE2ETestStep) Run(ctx context.Context, dry bool) error {
 		return nil
 	}
 	var errs []error
-	if err := s.runSteps(ctx, s.pre); err != nil {
+	var err error
+	if err = s.runSteps(ctx, s.pre, ""); err != nil {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: Don't we need the secret in the pre-steps so the installer can commit the KUBECONFIG to it in the first place?

File: pkg/steps/new_e2e.go [hunk 17]
```diff
@@ -181,6 +199,50 @@ func imageForStep(image string, params api.Parameters) (string, error) {
...
+	if os.IsNotExist(err) {
+		return "", nil
+	} else if err != nil {
+		return "", err
+	}
+	secret := coreapi.Secret{
+		ObjectMeta: meta.ObjectMeta{Name: "installer"},
+		Data:       make(map[string][]byte, len(files)),
+	}
+	for _, f := range files {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: This seems fragile -- if the secret is write-able (or if we use a configmap) the pre-steps can just self-manage this

File: pkg/steps/new_e2e.go [hunk 18]
```diff
@@ -265,3 +271,73 @@ func (s newE2ETestStep) runPods(ctx context.Context, pods []coreapi.Pod, notifie
 	}
 	return nil
 }
+
+type openshiftInstallerTestStep struct {
```
Comment (stevekuznetsov) [inline pkg/steps/new_e2e.go]: I don't understand what is going on in this commit

File: pkg/api/config.go [hunk 19]
```diff
@@ -110,8 +110,9 @@ func validateTestStepConfiguration(fieldRoot string, input []TestStepConfigurati
 			validationErrors = append(validationErrors, fmt.Errorf("%s[%d].as: '%s' is not valid value, should be [a-zA-Z0-9_.-]", fieldRoot, num, test.As))
 		}
 
-		if len(test.Commands) == 0 {
-			validationErrors = append(validationErrors, fmt.Errorf("%s[%d].commands: is required", fieldRoot, num))
+		hasCommands, hasSteps := len(test.Commands) != 0, len(test.Steps) != 0
+		if !(hasCommands || hasSteps) || (hasCommands && hasSteps) {
+			validationErrors = append(validationErrors, fmt.Errorf("%s[%d]: either `commands` or `steps` (but not both) should be set", fieldRoot, num))
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/api/config.go]: Separate error messages are fine with me.

As for validation in general, as you said below, it is very incomplete. This was mostly a reminder to do it, I didn't want to spend too much time writing validation code  because I expect the configuration format to change.

File: pkg/api/config.go [hunk 20]
```diff
@@ -110,8 +110,9 @@ func validateTestStepConfiguration(fieldRoot string, input []TestStepConfigurati
 			validationErrors = append(validationErrors, fmt.Errorf("%s[%d].as: '%s' is not valid value, should be [a-zA-Z0-9_.-]", fieldRoot, num, test.As))
 		}
 
-		if len(test.Commands) == 0 {
-			validationErrors = append(validationErrors, fmt.Errorf("%s[%d].commands: is required", fieldRoot, num))
+		hasCommands, hasSteps := len(test.Commands) != 0, len(test.Steps) != 0
+		if !(hasCommands || hasSteps) || (hasCommands && hasSteps) {
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/api/config.go]: Do you mean the actual `^` operator? That is how I would do it in a language where booleans are `int`s, but I could not find find a way to do it in go.

File: pkg/api/config.go [hunk 21]
```diff
@@ -186,7 +187,7 @@ func validateReleaseTagConfiguration(fieldRoot string, input ReleaseTagConfigura
 
 func validateClusterProfile(fieldRoot string, p ClusterProfile) []error {
 	switch p {
-	case ClusterProfileAWS, ClusterProfileAWSAtomic, ClusterProfileAWSCentos, ClusterProfileAWSCentos40, ClusterProfileAWSGluster, ClusterProfileAzure4, ClusterProfileGCP, ClusterProfileGCP40, ClusterProfileGCPHA, ClusterProfileGCPCRIO, ClusterProfileGCPLogging, ClusterProfileGCPLoggingJournald, ClusterProfileGCPLoggingJSONFile, ClusterProfileGCPLoggingCRIO, ClusterProfileOpenStack, ClusterProfileVSphere:
+	case "", ClusterProfileAWS, ClusterProfileAWSAtomic, ClusterProfileAWSCentos, ClusterProfileAWSCentos40, ClusterProfileAWSGluster, ClusterProfileAzure4, ClusterProfileGCP, ClusterProfileGCP40, ClusterProfileGCPHA, ClusterProfileGCPCRIO, ClusterProfileGCPLogging, ClusterProfileGCPLoggingJournald, ClusterProfileGCPLoggingJSONFile, ClusterProfileGCPLoggingCRIO, ClusterProfileOpenStack, ClusterProfileVSphere:
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/api/config.go]: Mostly a remnant from the idea of having an empty value be "choose one randomly", which is what I used initially to trigger the new code path.

I'll remove it, It has no place in this PR.

File: pkg/defaults/defaults.go [hunk 22]
```diff
@@ -415,7 +420,9 @@ func stepConfigsForBuild(config *api.ReleaseBuildConfiguration, jobSpec *api.Job
 		switch {
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/defaults/defaults.go]: In my defense, I only extended the nonsense =)

The change to the last case is also not necessary anymore.

File: pkg/steps/new_e2e.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+		name:      config.As,
+		podClient: podClient,
+		jobSpec:   jobSpec,
+		pre:       config.PreSteps,
+		tests:     config.Steps,
+		post:      config.PostSteps,
+	}
+}
+
+func (s *newE2ETestStep) Inputs(ctx context.Context, dry bool) (api.InputDefinition, error) {
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: That is added in a subsequent commit, as you probably saw after commenting, but even then it is not correct. Input resolution happens before we know which images come from `release_tag`, so it won't work.

Note that the same mechanism is used in `imageForStep`, but by then we have already imported the images. I also got strange results when I tested this where the variables were not added to `params` because the IS seemed to be empty. I'll have to investigate this.

File: pkg/steps/new_e2e.go [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+			errs = append(errs, fmt.Errorf("%q test steps failed: %v", s.name, err))
+		}
+	}
+	if err := s.runSteps(ctx, s.post); err != nil {
+		errs = append(errs, fmt.Errorf("%q post steps failed: %v", s.name, err))
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (s *newE2ETestStep) Done() (bool, error) { return false, nil }
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: TODO. Tests pass without it =)

File: pkg/steps/new_e2e.go [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+func (s *newE2ETestStep) Name() string        { return s.name }
+func (s *newE2ETestStep) Description() string {
+	return fmt.Sprintf("Run cluster install %s", s.name)
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: Yeah, so is naming of types and files and all the rest. I'll polish when the implementation is solid.

File: pkg/steps/new_e2e.go [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+		ret = append(ret, api.InternalImageLink(api.PipelineImageStreamTagReference(step.Image)))
+	}
+	return
+}
+
+func (s *newE2ETestStep) Creates() []api.StepLink { return nil }
+func (s *newE2ETestStep) Provides() (api.ParameterMap, api.StepLink) {
+	return nil, nil
+}
+func (s *newE2ETestStep) SubTests() []*junit.TestCase { return nil }
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: Yes, everything related to junit is TODO.

File: pkg/steps/new_e2e.go [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+			continue
+		}
+		name := s.name + "-" + step.Name
+		pod, err := generateBasePod(s.jobSpec, resources)
+		pod.ObjectMeta.Name = name
+		pod.ObjectMeta.Annotations[annotationContainersForSubTestResults] = step.Name
+		container := &pod.Spec.Containers[0]
+		container.Name = step.Name
+		container.Image = step.Image
+		container.Command = []string{"/bin/sh", "-c", "#!/bin/sh\nset -eu\n" + step.Commands}
```
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: I copied this from `pod.go`, which uses `sh`. Later on I found that the template I used as reference requires `bash`, which is why it is changed in the last commit.

We should change `pod.go` to also use `bash`, should be backwards compatible.

File: pkg/steps/new_e2e.go (hunk 1 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: Yes, the branch I linked above is based on https://github.com/openshift/release/commit/dc0e476cbb24354a30a58694687657b143aad698, which is a few weeks old now, but I looked briefly at the diff and didn't see anything major.

I really don't want to chase the templates, so I'll keep monitoring them for big changes but will only pull the changes when we are ready.

File: pkg/steps/new_e2e.go (hunk 3 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: I debated whether `post` should short circuit or if the steps should `exit 0`. I went for the latter only because it made the implementation cleaner.

I believe the templates don't fail if `teardown` fails.

File: pkg/steps/new_e2e_test.go (hunk 5 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e_test.go]: Because `steps[0].Image` should have been just `image0`. Bad test.

File: pkg/api/types.go (hunk 6 above)
Comment (bbguimaraes) [reply pkg/api/types.go]: Because the granularity there is per-test, this allows for different resource requirements for each step. I don't really understand why it is a separate top-level field instead of being a field in `TestStepConfiguration`, but we can extend it if there is a good reason.

File: pkg/steps/new_e2e.go (hunk 8 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: We don't even need the `struct` because we know what `podName` will be based on `containerName`. I also want to check if we can use a single worker for all three phases.

File: pkg/steps/new_e2e.go (hunk 9 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: "Unset" here means the flag wasn't present in the command line, which is a normal mode of operation. A more general question is: should we extract the information we need via artifacts, which I'll address below…

File: pkg/steps/new_e2e_test.go (hunk 10 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e_test.go]: Yes, tests are by no means comprehensive yet.

File: pkg/steps/new_e2e_test.go (hunk 11 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e_test.go]: Because the step doesn't have an `ArtifactDir`.

File: pkg/steps/new_e2e.go (hunk 12 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: I am having a really hard time with naming, suggestions highly welcome.

File: pkg/steps/new_e2e.go (hunk 13 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: Artifact from a previous implementation, all `err`s can be scoped now.

File: pkg/steps/new_e2e.go (hunk 15 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: Depends if we want `post` if `pre` failed.

File: pkg/steps/new_e2e.go (hunk 16 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: The main reason for repurposing artifacts is because the `installer` image doesn't have `oc` in it. If we can enforce that `pre` steps must be able to modify a secret, I'll gladly remove this.

File: pkg/steps/new_e2e.go (hunk 18 above)
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: First of all, in the best scenario it is intended to be removed. Up to this point, the implementation is completely generic and has no E2E- (or even Openshift-) specific bits.

The code in `openshiftInstallerTestStep` is everything I still don't know how to make generic that is required for the E2E AWS test.. It is also a possible implementation where different test types would share the step back-end but have test-specific extensions, but I would prefer if we could make all that part of configuration / step implementation.

File: pkg/api/config.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/api/config.go]: I was being silly :)

File: pkg/steps/new_e2e.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/steps/new_e2e.go]: I don't think the job will finish without it, though

File: pkg/steps/new_e2e.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/steps/new_e2e.go]: I'm amazed `pod.go` also uses `sh` and this hasn't broken anyone yet

[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/steps/new_e2e.go]: Let's update that in a separate PR

File: pkg/steps/new_e2e.go (hunk 3 above)
Comment (stevekuznetsov) [reply pkg/steps/new_e2e.go]: It's more of a question of what API we want to expose to users here

File: pkg/steps/new_e2e.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: what_if_I_told_you.png

File: pkg/steps/new_e2e.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (bbguimaraes) [reply pkg/steps/new_e2e.go]: https://github.com/openshift/ci-tools/pull/87

File: cmd/ci-operator/main.go (hunk 7 above)
Comment (bbguimaraes) [reply cmd/ci-operator/main.go]: I agreed initially, but then realized we would need code here anyway to validate that the key and the `name` field match (or similar treatment) and that duplicates would be silently ignored. Loading a list and validating that names are not duplicated seems better.

File: pkg/steps/new_e2e.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/steps/new_e2e.go]: oh no
